cpp/src/arrow/c/bridge.h

```diff
@@ -0,0 +1,163 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "arrow/c/abi.h"
+#include "arrow/result.h"
+#include "arrow/status.h"
+#include "arrow/type_fwd.h"
+#include "arrow/util/macros.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+/// \brief Export C++ DataType using the C data interface format.
+///
+/// The root type is considered to have empty name and metadata.
+/// If you want the root type to have a name and/or metadata, pass
+/// a Field instead.
+///
+/// \param[in] type DataType object to export
+/// \param[out] out C struct where to export the datatype
+ARROW_EXPORT
+Status ExportType(const DataType& type, struct ArrowSchema* out);
```

Comment:
Hmm, no, the `ArrowSchema` pointer is given by the caller, we don't want C++ to move it around.
(it may actually work to return `Result`, but I don't want to be constrained by that)

Reply:
This is an internal API so I won't block, but FTR I don't think it'd be constraining:
```diff
  ArrowSchema* my_schema = // ...
- RETURN_NOT_OK(ExportType(some_type, my_schema));
+ ARROW_ASSIGN_OR_RAISE(*my_schema, ExportType(some_type));
```

Reply:
The problem is: what happens when C++ moves the C struct around?

Reply:
Nothing; the C struct is relocatable by design- otherwise `ArrowSchemaMove` wouldn't work.

Reply:
Hmm. I'm not fond of it in any case. The current signature mimicks the intended usage in C land, so in this case I think the pointer-out mechanism is actually beneficial.